lambdas/preview/index.py

```diff
@@ -141,35 +145,102 @@ def extract_vcf(file_):
-            # don't escape quotes
-            line = line.strip().decode('unicode_escape')
-            formatted = f'<p>{line}</p>'
+            line = _truncate(line.rstrip().decode(), size)
             if line.startswith('##'):
-                meta.append(formatted)
+                meta.append(line)
             elif line.startswith('#'):
-                header.append(formatted)
+                header.append(line)
             else:
-                data.append(formatted)
+                data.append(line)
+            size += len(line)
+            # stop if we're over the max
+            if index >= MAX_LINES or size > MAX_BYTES:
+                break
 
-            if index >= MAX_LINES:
+    info = {
+        'data': {
+            'meta': meta,
+            'header': header,
+            'data': data,
+        }
+    }
+
+    return '', info
+
+def extract_txt(file_):
+    """
+    Display first and last few lines of a potentially large file.
+
+    Args:
+        file_ - file-like object
+    """
+    size = 0
+    ellipsis = False
+    head = []
+    with open(file_.name, 'rb') as txt:
+        for index, line in enumerate(txt, start=1):
+            line = _truncate(line.rstrip().decode(), size)
+            size += len(line)
+            # this is a heuristic; can fail to return shorter tail lines
+            # if head lines are uncharacteristically long
+            # we're guarding against, for example, huge single-line JSON files
+            if index <= MAX_LINES and size <= MAX_BYTES:
+                head.append(line)
+            if size > MAX_BYTES:
                 break
+            if index > MAX_LINES:
+                ellipsis = True
+                break
+    tail = []
+    if ellipsis: # in this case we need a tail
+        headlen = int(MAX_LINES/2) # pylint: disable=old-division
+        taillen = MAX_LINES - headlen
+        # cut head
+        head = head[:headlen]
+        # grow tail
+        with open(file_.name, 'rb') as txt:
+            count = 0
+            txt.seek(0, os.SEEK_END) # the very end
+            while count < taillen:
+                txt.seek(-2, os.SEEK_CUR)
+                if txt.read(1) == b'\n':
+                    count +=1 
+
+            for _ in range(taillen):
+                line = txt.readline().rstrip().decode()
+                size += len(line)
+                if size < MAX_BYTES:
+                    tail.append(line)
+
+    info = {
+        'data': {
+            'head': head,
+            'tail': tail
+        }
+    }
+    return '', info
 
-    # pylint: disable=bad-continuation
-    html = ("<div>"
-                "<div data-type='meta'>{}</div>"
-                "<div data-type='header'>{}</div>"
-                "<div data-type='data'>{}</div>"
-            "</div>").format(
-                ''.join(meta) + '<br />',
-                ''.join(header) + '<br />',
-                ''.join(data))
 
-    return html, {}
+def _truncate(line, size):
```

Comment:
This function looks weird to me. I'd rather pass in the remaining bytes. Then, the whole function could just be:
```
return line[:max(remaining, 0)]
```